pkg/bgp/speaker.go

```diff
@@ -0,0 +1,96 @@
+package bgp
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	bgpconfig "github.com/cilium/cilium/pkg/bgp/config"
+	bgpk8s "github.com/cilium/cilium/pkg/bgp/k8s"
+	bgplog "github.com/cilium/cilium/pkg/bgp/log"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	nodetypes "github.com/cilium/cilium/pkg/node/types"
+	"github.com/cilium/cilium/pkg/option"
+	"go.universe.tf/metallb/pkg/k8s/types"
+	metallbspr "go.universe.tf/metallb/pkg/speaker"
+)
+
+// Speaker provides a method set for interfacing
+// with a BGP speaker.
+//
+// This interface is heavily modeled after MetalLB's speaker
+// as it's the first BGP integration for Cilium's use cases.
+//
+// If other BGP integrations are desired, consider building out custom types
+// and a more abstracted method set.
+type Speaker interface {
+	SetService(name string, svc *metallbspr.Service, eps *metallbspr.Endpoints) types.SyncState
+	SetNodeLabels(labels map[string]string) types.SyncState
+	PeerSessions() []metallbspr.Session
+}
+
+// MetalLBSpeaker implements the bgp.Speaker interface
+// and is thin wrapper around the metallb controller proper.
+type MetalLBSpeaker struct {
+	C      *metallbspr.Controller
+	logger *bgplog.Logger
+}
+
+// NewMetalLBSpeaker will create a new Speaker powered by
+// a MetalLB BGP Speaker.
+//
+// This constructor expects option.Config.BGPConfigPath to point to
+// a valid filesystem path where a MetalLB configure resides.
+// It's an error if the config cannot be parsed.
+//
+// The MetalLB speaker will use the value of nodetypes.GetName() as
+// its node identity.
+func NewMetalLBSpeaker(ctx context.Context) (Speaker, error) {
+	var (
+		log = logging.DefaultLogger.WithField(logfields.LogSubsys, "bgp-speaker")
+	)
+
+	logger := &bgplog.Logger{Entry: log}
+	client := bgpk8s.New(logger.Logger)
+
+	c, err := metallbspr.NewController(metallbspr.ControllerConfig{
+		MyNode:        nodetypes.GetName(),
+		Logger:        logger,
+		SList:         nil, // BGP speaker doesn't use speakerlist
+		DisableLayer2: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	c.Client = client
+
+	f, err := os.Open(option.Config.BGPConfigPath)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	config, err := bgpconfig.Parse(f)
+	if err != nil {
+		return nil, err
+	}
+	if ss := c.SetConfig(logger, config); ss == types.SyncStateError {
+		return nil, fmt.Errorf("Setting MetalLB config failed. View logs for details")
+	}
+
+	return MetalLBSpeaker{
+		C:      c,
+		logger: logger,
+	}, nil
+}
+
+func (m MetalLBSpeaker) SetService(name string, svc *metallbspr.Service, eps *metallbspr.Endpoints) types.SyncState {
+	return m.C.SetService(m.logger, name, svc, eps)
+}
+func (m MetalLBSpeaker) SetNodeLabels(labels map[string]string) types.SyncState {
+	return m.C.SetNodeLabels(m.logger, labels)
+}
+func (m MetalLBSpeaker) PeerSessions() []metallbspr.Session {
+	return m.C.PeerSessions()
```

Comment:
These should probably be pointers because when creating an object to adhere to the `Speaker` interface, it's going to need to be a pointer anyway. Any reason why they weren't initially?

```suggestion
func (m *MetalLBSpeaker) SetService(name string, svc *metallbspr.Service, eps *metallbspr.Endpoints) types.SyncState {
	return m.C.SetService(m.logger, name, svc, eps)
}
func (m *MetalLBSpeaker) SetNodeLabels(labels map[string]string) types.SyncState {
	return m.C.SetNodeLabels(m.logger, labels)
}
func (m *MetalLBSpeaker) PeerSessions() []metallbspr.Session {
	return m.C.PeerSessions()
```

Reply:
Typically, when all members of a struct are pointers, I pass the container object by value. The container struct doesn't need to add to gc pressure (by being a heap allocation on every construction). 

I typically default to copying over passing by reference until i can measure and know copying is a problem. 

Im not immediately seeing the issue with the receiver being a pointer or a value, since the constructor returns the interface type.